pennylane/pulse/convenience_functions.py

```diff
@@ -0,0 +1,117 @@
+
+    **Example**
+
+    The ``constant`` function can be used to create a parametrized hamiltonian
+
+    >>> windows = [(1, 7), (9, 14)]
+    >>> H = qml.pulse.constant(windows) * qml.PauliX(0)
+
+    When calling the parametrized hamiltonian, ``constant`` will return the input parameter only
+    when the time is inside the given time windows
+
+    >>> params = [5]
+    >>> H(params, t=8)  # t is outside the time windows
+    0.0*(PauliX(wires=[0]))
+    >>> H(params, t=5)  # t is inside the time windows
+    5.0*(PauliX(wires=[0]))
+    """
+    return rect(x=lambda p, _: p, windows=windows)
+
+
+def rect(x: Union[float, Callable], windows: List[Tuple[float]] = None):
+    """Multiplies ``x`` by a rectangular function, returning a callable ``f(p, t)`` that evaluates
+    the given function/scalar ``x`` inside the time windows defined in ``windows``.
+
+    .. note::
+
+        If ``x`` is a function, it must accepts two arguments: the trainable parameters and time.
+
+    Args:
+        x (Union[float, Callable]): a scalar or a function that accepts two arguments: the trainable
+            parameters and time
+        windows (Tuple[float, Tuple[float]]): List of tuples containing time windows where x is
+            evaluated. If ``None`` it is always evaluated. Defaults to ``None``.
+
+    **Example**
+
+    The ``rect`` function can be used to create a parametrized hamiltonian
+
+    >>> def f1(p, t):
+    ...     return jnp.polyval(p, t)
+    >>> windows = [(1, 7), (9, 14)]
+    >>> H = qml.pulse.rect(f1, windows) * qml.PauliX(0)
+
+    When calling the parametrized hamiltonian, ``rect`` will evaluate the given function only
+    inside the time windows
+
+    >>> params = [jnp.ones(4)]
+    >>> H(params, t=8)  # t is outside the time windows
+    0.0*(PauliX(wires=[0]))
+    >>> H(params, t=5)  # t is inside the time windows
+    156.0*(PauliX(wires=[0]))
+
+    One can also pass a scalar to the ``rect`` function
+
+    >>> H = qml.pulse.rect(10, windows) * qml.PauliX(0)
+
+    In this case, ``rect`` will return the given scalar only when the time is inside the provided
+    time windows
+
+    >>> params = [None]  # the parameter value won't be used!
+    >>> H(params, t=8)
+    0.0*(PauliX(wires=[0]))
+    >>> H(params, t=5)
+    10.0*(PauliX(wires=[0]))
+    """
+    if not has_jax:
+        raise ImportError(
+            "Module jax is required for any pulse-related convenience function. "
+            "You can install jax via: pip install jax"
+        )
+    if not callable(x):
+
+        def _f(_, __):
+            return jnp.array(x, dtype=float)
+
+    else:
+        _f = x
+
+    def f(p, t):
+        p = jnp.array(p, dtype=float)  # if p is an integer, f(p, t) will be cast to an integer
+        if windows is not None:
+            return jnp.where(
+                jnp.any(jnp.array([(t >= ti) & (t <= tf) for ti, tf in windows])), _f(p, t), 0
+            )
```

Comment:
@AlbertMitjans  you moved past me too quickly and I had not a chance to comment about this:

this code is (probably) jax-suboptimal because you are unpacking an array into several scalars. 
Best case scenario this slows down compilation. worst case runtime as well.

I would rewrite this as follows:
 
```python
        if windows is not None:
            ti, tf = zip(*windows)
            ti, tf = jnp.array(ti), jnp.array(tf)
            return jnp.where(
                jnp.any((t >= ti) & (t <= tf)), _f(p, t), 0
            )
```

Reply:
Hi @PhilipVinc ! Thanks for the suggestion! Opened a PR adding this implementation: https://github.com/PennyLaneAI/pennylane/pull/3695